plugins/janus_sip.c

```diff
@@ -4656,18 +4659,35 @@ static void *janus_sip_handler(void *data) {
 					session->stack->s_nh_m = nua_handle(session->stack->s_nua, session, TAG_END());
 				}
 				janus_mutex_unlock(&session->stack->smutex);
+				/* call_id is set to be later used in the delivery recepit */
+				json_t *request_callid = json_object_get(root, "call_id");
+				/* Take call-id from request, if it exists */
+				if(request_callid) {
+					message_callid = g_strdup(json_string_value(request_callid));
+				} else {
+					/* If call-id does not exist in request, create a random one */
+					message_callid = g_malloc0(24);
+					janus_sip_random_string(24, message_callid);
+				}
 				nua_message(session->stack->s_nh_m,
 					SIPTAG_TO_STR(uri_text),
 					SIPTAG_CONTENT_TYPE_STR(content_type),
 					SIPTAG_PAYLOAD_STR(msg_content),
 					NUTAG_PROXY(session->helper && session->master ?
 						session->master->account.outbound_proxy : session->account.outbound_proxy),
 					TAG_IF(strlen(custom_headers) > 0, SIPTAG_HEADER_STR(custom_headers)),
+					SIPTAG_CALL_ID_STR(message_callid),
 					TAG_END());
 			}
 			/* Notify the operation */
 			result = json_object();
 			json_object_set_new(result, "event", json_string("messagesent"));
+			/* message_callid will be later used to return the delivery status of the message */
+			json_object_set_new(result, "call_id", json_string(message_callid));
+			if (session->callid == NULL) {
+				session->callid = g_strdup(message_callid);
```

Comment:
I see your point, in most implementation I've seen the SIP MESSAGEs are exchanged outside of a dialog even when a call is active between the peers. I will research a bit more but I think an hashtable for the SIP MESSAGEs may be unnecessary. What would be your preference?

Reply:
Sorry, missed this message before.

One one end, I think you maybe don't need to track it at all, since when you get a response in `nua_r_message` you just read the Call-ID from the message you received, and pass that back as an event. On the other end, though, the way you're sending the event right now won't always work when using session helpers, as you may be sending a message from helper n.2, but the delivered event would be sent to the session master (the session owning the loop).

As such, there may indeed be a need to keep a hashtable in the session, or even just a list: this way, when you get a `nua_r_message` back, you start from the master session (the one the loop notifies about) and try to find the message Call-ID there; if it can't be found, then iterate on helper sessions until you find the Call-ID, and if you find it that's the session you send the event to.

Of course, if there's a master session and helper sessions, they all belong to the same application, so we may even just keep it simple and say that, whatever session sent the message, the delivered event will always be notified back to the master session. In that case, nothing else you need to do, except maybe add some text in the documentation to explain this and get rid of the

			if (session->callid == NULL) {
				session->callid = g_strdup(message_callid);

Reply:
> If my understanding is correct Helpers are used to offer concurrency for sip calls on the same account, something that IMHO is not needed for messages as messages live and die within the server response (success or error).

That depends. If I'm using helpers to have concurrent calls with you (master session), Alice (helper 1) and Bob (helper 2), sending an in-dialog MESSAGE to Alice will happen on a helper session, but the "delivered" event would be sent to the master session, which may be seen as incorrect from a signalling flow perspective. This can be problematic depending how the application is written, which is why I said that if we say this is ok, it needs to be documented, because it means that it's up to the application (e.g., the JavaScript code) to keep a global map of MESSAGE session IDs to know which call the acknowledgement actually refers to.

> Additionally, on a personal note, I am not too confident I can perform such a structural change as my skills have dropped a lot since I've last developed in C. C is not like riding a bicycle (ahah).

That's absolutely fine, we can indeed revisit this later (after it's merged) if it turns out to be a problem in practice.